security/pkg/util/vaultutil.go

```diff
@@ -0,0 +1,277 @@
+	resp, err := client.Logical().Write(
+		loginPath,
+		map[string]interface{}{
+			"jwt":  sa,
+			"role": role,
+		})
+
+	if err != nil {
+		return "", err
+	}
+	return resp.Auth.ClientToken, nil
+}
+
+// SignCsrByVault signs the CSR and return the signed certifcate and the CA certificate chain
+// Return the signed certificate and the CA certificate chain when succeed.
+// client: the Vault client
+// csrSigningPath: the path for signing a CSR
+// csr: the CSR to be signed, in pem format
+func SignCsrByVault(client *api.Client, csrSigningPath string, csr []byte) ([]byte, []byte, error) {
+	m := map[string]interface{}{
+		"format": "pem",
+		"csr":    string(csr[:]),
+	}
+	res, err := client.Logical().Write(csrSigningPath, m)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to post to %v: %v", csrSigningPath, err)
+	}
+	//Extract the certificate and the certificate chain
+	certificate, ok := res.Data["certificate"]
+	if !ok {
+		return nil, nil, fmt.Errorf("no certificate in the CSR response")
+	}
+	cert, ok := certificate.(string)
+	if !ok {
+		return nil, nil, fmt.Errorf("the certificate in the CSR response is not a string")
+	}
+	caChain, ok := res.Data["ca_chain"]
+	if !ok {
+		return nil, nil, fmt.Errorf("no certificate chain in the CSR response")
+	}
+	chain, ok := caChain.([]interface{})
+	if !ok {
+		return nil, nil, fmt.Errorf("the certificate chain in the CSR response is of unexpected format")
+	}
+	certChain := ""
+	for i, c := range chain {
+		_, ok := c.(string)
+		if !ok {
+			return nil, nil, fmt.Errorf("the certificate in the certificate chain is not a string")
+		}
+		certChain += c.(string)
+		if i < len(chain) {
+			certChain += "\n"
+		}
+	}
+
+	return []byte(cert), []byte(certChain), nil
+}
+
+// ReviewServiceAccountAtK8sAPIServer reviews the CSR credential (k8s service account) at k8s API server.
+// k8sAPIServerURL: the URL of k8s API Server
+// k8sAPIServerCaCert: the CA certificate of k8s API Server
+// reviewerToken: the service account of the k8s token reviewer
+// csrCredential: the credential of the CSR requester
+func ReviewServiceAccountAtK8sAPIServer(k8sAPIServerURL string, k8sAPIServerCaCert []byte,
+	reviewerToken string, csrCredential []byte) (*http.Response, error) {
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(k8sAPIServerCaCert)
+	saReq := saValidationRequest{
+		APIVersion: "authentication.k8s.io/v1",
+		Kind:       "TokenReview",
+		Spec:       specForSaValidationRequest{Token: string(csrCredential[:])},
+	}
+	saReqJSON, err := json.Marshal(saReq)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal the service account review request: %v", err)
+	}
+	req, err := http.NewRequest("POST", k8sAPIServerURL, bytes.NewBuffer(saReqJSON))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a HTTP request: %v", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", "Bearer "+reviewerToken)
```

Comment:
Is this part of the call credential to let ApiServer accept your request?

Reply:
If so, this is also a jwt, right?

Reply:
The answer is yes to both of your questions.